src/py/flwr/server/superlink/driver/serverappio_servicer.py

```diff
@@ -215,6 +215,15 @@ def PullMessages(
         messages_list = []
         while messages_res:
             msg = messages_res.pop(0)
+
+            # Set the `run_id` of the Run if the message was created
+            # by the SuperLink itself as a reply Message that expired
+            if (
+                msg.metadata.src_node_id == SUPERLINK_NODE_ID
+                and msg.metadata.dst_node_id == SUPERLINK_NODE_ID
+            ):
+                msg.metadata._run_id = request.run_id
+
             _raise_if(
                 validation_error=request.run_id != msg.metadata.run_id,
                 request_name="PullMessages",
```

Comment:
Okay, I think this should be my final thought. It's better not to alter the automatically generated Messages, and just skip the extra check here.

```suggestion

            # Skip `run_id` check for SuperLink generated replies
            if msg.metadata.src_node_id != SUPERLINK_NODE_ID:
                _raise_if(
                    validation_error=request.run_id != msg.metadata.run_id,
                    request_name="PullMessages",
```

Reply:
much cleaner

Reply:
Instead of modifying the message later or skipping the check here, can we add then `run_id` directly when we create the message?